src/poprox_recommender/components/embedders/topic_wise_user.py

```diff
@@ -92,61 +96,187 @@
         external_id=topic,
         raw_data={},
     )
-    for topic, description in topic_descriptions.items()
+    for topic, description in TOPIC_DESCRIPTIONS.items()
 ]
 
 
-def virtual_clicks(onboarding_topics, topic_embeddings):
+def virtual_clicks(onboarding_topics, topic_articles):
+    topic_uuids_by_name = {article.external_id: article.article_id for article in topic_articles}
     virtual_clicks = []
     for interest in onboarding_topics:
         topic_name = interest.entity_name
-        preference = interest.preference or 0
+        preference = interest.preference or 1
 
-        if topic_name in topic_embeddings:
-            entity_id = interest.entity_id
+        if topic_name in topic_uuids_by_name:
+            article_id = topic_uuids_by_name[topic_name]
 
-            virtual_clicks.extend([Click(article_id=entity_id)] * preference)
+            virtual_clicks.extend([Click(article_id=article_id)] * (preference - 1))
     return virtual_clicks
 
 
-class TopicUserEmbedder(NRMSUserEmbedder):
+class UserOnboardingEmbedder(NRMSUserEmbedder):
     article_embedder: NRMSArticleEmbedder
     embedded_topic_articles: ArticleSet | None = None
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, embedding_source: str = "static", topic_embedding: str = "nrms", **kwargs):
         super().__init__(*args, **kwargs)
         self.article_embedder = NRMSArticleEmbedder(
             model_file_path("nrms-mind/news_encoder.safetensors"), device=self.device
         )
+        self.embedding_source = embedding_source
+        self.topic_embedding = topic_embedding
 
     @torch_inference
-    def __call__(self, clicked_articles: ArticleSet, interest_profile: InterestProfile) -> InterestProfile:
-        if len(clicked_articles.articles) == 0:
-            interest_profile.embedding = None
-        else:
-            if self.embedded_topic_articles is None:
-                self.embedded_topic_articles = self.article_embedder(ArticleSet(articles=topic_articles))
-            topic_embeddings_by_name = {
-                article.external_id: embedding
-                for article, embedding in zip(topic_articles, self.embedded_topic_articles.embeddings)
-            }
+    def __call__(
+        self, candidate_articles: ArticleSet, clicked_articles: ArticleSet, interest_profile: InterestProfile
+    ) -> InterestProfile:
+        if self.embedded_topic_articles is None:
+            self.embedded_topic_articles = self.article_embedder(ArticleSet(articles=TOPIC_ARTICLES))
+
+        topic_embeddings_by_uuid = {
+            article.article_id: embedding
+            for article, embedding in zip(TOPIC_ARTICLES, self.embedded_topic_articles.embeddings)
+        }
+
+        topic_clicks = virtual_clicks(interest_profile.onboarding_topics, TOPIC_ARTICLES)
+
+        embeddings_from_definitions = self.build_embeddings_from_definitions()

```

Comment:
Here's the problem. This is re-run every call, instead of at setup, which looks unnecessary. Further, `build_embeddings_from_definitions` instantiates a model on every call. The result is that every recommendation request re-instantiates an embedding model.

The fix for the time being is to move either the instantiation, or the entire definition embedding, into `__init__`.

The fix in the future (once component caches exist) is to run the topics through the same embedding component as we use for everything else :).